WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Package Selection/Package Creation/ShippingLabelAddNewPackageViewModel.swift

```diff
@@ -5,23 +5,157 @@ import Yosemite
+                            _ servicePackage: ShippingLabelPredefinedPackage? ,
+                            _ packagesResponse: ShippingLabelPackagesResponse?) -> Void
+    let onCompletion: Completion
 
     init(_ selectedIndex: Int = PackageViewType.customPackage.rawValue,
-         packagesResponse: ShippingLabelPackagesResponse?) {
+         stores: StoresManager = ServiceLocator.stores,
+         siteID: Int64,
+         packagesResponse: ShippingLabelPackagesResponse?,
+         onCompletion: @escaping Completion) {
         self.selectedIndex = selectedIndex
-        self.servicePackageVM = ShippingLabelServicePackageListViewModel(packagesResponse: packagesResponse)
+        self.stores = stores
+        self.siteID = siteID
+        self.packagesResponse = packagesResponse
+        self.onCompletion = onCompletion
     }
 
     enum PackageViewType: Int {
         case customPackage = 0
         case servicePackage = 1
     }
 }
+
+// MARK: - Helper methods
+extension ShippingLabelAddNewPackageViewModel {
+
+    /// Creates the custom package remotely and updates the package details to select the new package
+    ///
+    func createCustomPackage(onCompletion: @escaping (Bool) -> Void) {
+        guard let newCustomPackage = validatedCustomPackage else {
+            onCompletion(false)
+            return
+        }
+
+        createPackage(customPackage: newCustomPackage) { [weak self] success in
+            onCompletion(success)
+
+            // On success, reset tab state and save new package details
+            guard let self = self, success else { return }
+            self.customPackageVM = ShippingLabelCustomPackageFormViewModel()
+            self.onCompletion(newCustomPackage, nil, self.packagesResponse)
+        }
+    }
+
+    /// Activates the selected service package remotely and updates the package details to select the new package
+    ///
+    func activateServicePackage(onCompletion: @escaping (Bool) -> Void) {
+        guard let selectedServicePackage = selectedServicePackage,
+              let shippingProvider = servicePackageVM.predefinedOptions.first(where: { $0.predefinedPackages.contains(selectedServicePackage) } ) else {
+            onCompletion(false)
+            return
+        }
+
+        let selectedOption = ShippingLabelPredefinedOption(title: shippingProvider.title,
+                                                           providerID: shippingProvider.providerID,
+                                                           predefinedPackages: [selectedServicePackage])
+
+        createPackage(predefinedOption: selectedOption) { [weak self] success in
+            onCompletion(success)
+
+            // On success, reset tab state and save new package details
+            guard let self = self, success else { return }
+            self.customPackageVM = ShippingLabelCustomPackageFormViewModel()
+            self.servicePackageVM.packagesResponse = self.packagesResponse
```

Comment:
I'd recommend avoiding direct mutations of objects from outside. For instance, since the service package view model's package response is internal and mutable, there's a risk that it can be updated accidentally from outside. This can be updated with better access control as follows:
- Keep the service package view model's `packagesResponse` private and immutable.
- Make `servicePackageVM` a variable instead of a lazy one so that you can reset it with a new package response when needed.

Reply:
Thank you! That's good advice for me to keep in mind, and makes a lot of sense here.